### PR TITLE
S24 migrate components typescript pt2

### PR DESCRIPTION
**Pull request by dli505**
A continuation of https://github.com/gordon-cs/gordon-360-ui/pull/2290
**Changes made in PR:**
- finished migration of Components to Typescript
- files finished:
    - Header
    - Loader
    - Nav
    - NavButton
    - PaletteSwitcherDialog
    - PhotoCropper
    - Profile
    - Providers
    - PWAInstructions
    - Snackbar
    - SpreadsheetUploader

### CODE REVIEW

File: src/components/Profile/components/PersonalInfoList/components/UpdatePhoneDialog/index.tsx [hunk 1]
```diff
@@ -67,20 +74,22 @@ const UpdatePhone = () => {
 
 // From material ui website
 // https://material-ui.com/components/text-fields/#integration-with-3rd-party-input-libraries
-const phoneMaskUS = forwardRef((props, ref) => {
-  const { onChange, ...other } = props;
+const phoneMaskUS = forwardRef(
+  (props: { onChange: ({}) => void; name?: string } & InputBaseComponentProps, ref) => {
```
Comment (EjPlatzer) [inline src/components/Profile/components/PersonalInfoList/components/UpdatePhoneDialog/index.tsx:78]: This is an incorrect type for `onChange` - please specify the type more accurately.

File: src/components/Snackbar/index.tsx [hunk 2]
```diff
@@ -1,12 +1,31 @@
...
+import { Snackbar, SnackbarProps, Alert, AlertProps } from '@mui/material/';
 import {
   CheckCircleOutlineOutlined,
   ErrorOutline,
   InfoOutlined,
   ReportProblemOutlined,
 } from '@mui/icons-material/';
 
-const SimpleSnackbar = ({ open, text, severity, duration = 10000, onClose, ...otherProps }) => {
+export type severityType = 'error' | 'info' | 'success' | 'warning';
```
Comment (EjPlatzer) [inline src/components/Snackbar/index.tsx:9]: We should use the `AlertColor` type provided by MUI instead.

File: src/components/SpreadsheetUploader/index.tsx [hunk 3]
```diff
@@ -13,14 +13,26 @@ const acceptedTypes = [
   'text/csv',
 ];
 
-const displayCell = (cellData) => {
+const displayCell = (cellData: any) => {
```
Comment (EjPlatzer) [inline src/components/SpreadsheetUploader/index.tsx:16]: Better to use `unknown` here:
```suggestion
const displayCell = (cellData: unknown) => {
```

File: src/components/SpreadsheetUploader/index.tsx [hunk 4]
```diff
@@ -13,14 +13,26 @@ const acceptedTypes = [
...
   if (cellData instanceof Date) {
     return cellData.toLocaleString();
   } else {
     return cellData;
   }
 };
 
+type Props = {
+  open: boolean;
+  setOpen: ({}) => void;
```
Comment (EjPlatzer) [inline src/components/SpreadsheetUploader/index.tsx:26]: Again, this is an inaccurate type for this callback function. Please correctly type the parameter.

File: src/components/SpreadsheetUploader/index.tsx [hunk 5]
```diff
@@ -143,7 +155,7 @@ const SpreadsheetUploader = ({
         dropZone
       )}
       <GordonSnackbar
-        open={error}
+        open={error ? true : false}
```
Comment (EjPlatzer) [inline src/components/SpreadsheetUploader/index.tsx:158]:
```suggestion
        open={Boolean(error)}
```

File: src/components/EventList/index.tsx [hunk 6]
```diff
@@ -71,7 +71,7 @@ const EventList = ({ events }: Props) => {
         <List className="gc360_event_list" disablePadding>
           {events?.length < 1
             ? noEvents
-            : events.map((event) => <EventItem event={event} key={event.Event_ID} />)}
+            : events.map((event) => <EventItem {...event} key={event.Event_ID} />)}
```
[the remark this answers is not in the document]
Comment (dli505) [reply src/components/EventList/index.tsx:74]: I think something might have gotten messed up when I was trying to resolve merge conflicts, thanks for catching it!

File: src/components/PWAInstructions/index.tsx [hunk 7]
```diff
@@ -118,12 +118,15 @@ const devices = {
   },
 };
 
-const PWAInstructions = (props) => {
-  const [device, setDevice] = useState(null);
-  const [platform, setPlatform] = useState(null);
+const PWAInstructions = (props: any) => {
```
[the remark this answers is not in the document]
Comment (dli505) [reply src/components/PWAInstructions/index.tsx:121]: That was my bad, I was using it as a placeholder and forgot to change it. What type should be passed in? I haven't been able to figure it out.

[the remark this answers is not in the document]
Comment (EjPlatzer) [reply src/components/PWAInstructions/index.tsx:121]: If we look at where it's used in [`GuestWelcome`](https://github.com/gordon-cs/gordon-360-ui/blob/8aee5804ed033c687e1b61a42fa20b2a629892d8/src/views/Home/components/GuestWelcome/index.jsx#L112), we can see the props that are passed in. The only tricky one is `deferredPWAPrompt`, which appears to be an instance of a [`BeforeInstallPromptEvent`](https://developer.mozilla.org/en-US/docs/Web/API/BeforeInstallPromptEvent).

[the remark this answers is not in the document]
Comment (dli505) [reply src/components/PWAInstructions/index.tsx:121]: Should I type  `deferredPWAPrompt` as `typeof BeforeInstallPromptEvent` or something else?

[the remark this answers is not in the document]
Comment (EjPlatzer) [reply src/components/PWAInstructions/index.tsx:121]: It's type should be `BeforeInstallPromptEvent`, no `typeof` needed, I believe.

File: src/components/GordonDialogBox/index.tsx [hunk 8]
```diff
@@ -37,7 +37,7 @@ const GordonDialogBox = ({
       event.key === 'Enter' &&
       event.currentTarget.classList.contains('MuiDialog-root')
     ) {
-      buttonClicked(event);
+      buttonClicked!(event);
```
Comment (EjPlatzer) [inline src/components/GordonDialogBox/index.tsx:40]: A better approach would be to only call the function if the `buttonClicked` prop is not undefined:
```suggestion
      buttonClicked &&
      event.key === 'Enter' &&
      event.currentTarget.classList.contains('MuiDialog-root')
    ) {
      buttonClicked(event);
```

File: src/components/Loader/index.tsx [hunk 9]
```diff
@@ -5,12 +5,12 @@ type Props = CircularProgressProps & {
...
 
-const GordonLoader = ({ size = 100, color = 'primary', inline = false }: Props) => {
-  if (inline) return <CircularProgress size={size} color={color} />;
+const GordonLoader = ({ size = 100, color = 'primary', inline = false }) => {
+  if (inline) return <CircularProgress size={size} color={color as Props['color']} />;
   return (
     <Grid className={styles.gordon_loader} container justifyContent="center" alignItems="center">
       <Grid item>
-        <CircularProgress size={size} color={color} />
+        <CircularProgress size={size} color={color as Props['color']} />
```
Comment (EjPlatzer) [inline src/components/Loader/index.tsx:13]: This looks like a regression. We shouldn't need to override the type checking for `color`.

File: src/components/PWAInstructions/index.tsx [hunk 10]
```diff
@@ -118,12 +119,21 @@ const devices = {
   },
 };
 
-const PWAInstructions = (props) => {
-  const [device, setDevice] = useState(null);
-  const [platform, setPlatform] = useState(null);
+type props = {
+  open: boolean;
+  handleDisplay: () => void;
+  deferredPWAPrompt: typeof BeforeInstallPromptEvent;
```
Comment (EjPlatzer) [inline src/components/PWAInstructions/index.tsx:125]: This is not actually the correct `BeforeInstallPromptEvent`. Apparently, that event is non-standard, only supported in Chrome, and so the type is not officially available anywhere. Here is a good guide on how to add it properly: https://stackoverflow.com/a/67171375

File: src/components/PWAInstructions/index.tsx [hunk 11]
```diff
@@ -118,12 +119,21 @@ const devices = {
...
+};
+
+const PWAInstructions = ({ open, handleDisplay, deferredPWAPrompt }: props) => {
```
Comment (EjPlatzer) [inline src/components/PWAInstructions/index.tsx:128]: Looks like you updated this to destructure `props`, but didn't update the code which is still trying to use the `props` object that doesn't exist anymore.

File: src/components/PWAInstructions/index.tsx [hunk 12]
```diff
@@ -118,12 +119,21 @@ const devices = {
   },
 };
 
-const PWAInstructions = (props) => {
-  const [device, setDevice] = useState(null);
-  const [platform, setPlatform] = useState(null);
+type props = {
```
Comment (EjPlatzer) [inline src/components/PWAInstructions/index.tsx:122]: Type names should be PascalCase (i.e. start with an uppercase letter).
```suggestion
type Props = {
```

**Review by EjPlatzer — CHANGES_REQUESTED** [review]
Looking great! Just a few more things.

**Review by EjPlatzer — CHANGES_REQUESTED** [review]
This is coming along great! The only thing left is that there are build errors, which need to be addressed before we can merge.

Specifically, the following errors:
```
src/components/PhotoCropper/index.tsx(222,9): error TS2322: Type 'AlertColor | null' is not assignable to type '"error" | "info" | "success" | "warning"'.
  Type 'null' is not assignable to type '"error" | "info" | "success" | "warning"'.

src/components/Profile/components/MembershipsList/index.tsx(105,16): error TS2786: 'MembershipsList' cannot be used as a JSX component.
  Its return type 'Element | Element[]' is not a valid JSX element.
    Type 'Element[]' is missing the following properties from type 'ReactElement<any, any>': type, props, key

src/components/Profile/components/PersonalInfoList/components/UpdatePhoneDialog/index.tsx(94,51): error TS2353: Object literal may only specify known properties, and 'name' does not exist in type 'EventTarget'.

src/components/Profile/index.tsx(64,29): error TS2322: Type 'FacStaffProfileInfo' is not assignable to type '{ BuildingDescription: string; OnCampusDepartment: string; OnCampusRoom: string; OnCampusPhone: string; PersonType: string; office_hours: string; Mail_Location: string; Mail_Description: ReactElement<...>; }'.
  Types of property 'Mail_Description' are incompatible.
    Type 'string' is not assignable to type 'ReactElement<any, string | JSXElementConstructor<any>>'.

src/views/CoCurricularTranscript/guestTranscript.tsx([18](https://github.com/gordon-cs/gordon-360-ui/actions/runs/9892608760/job/27325688832?pr=2318#step:6:19)0,3): error TS2561: Object literal may only specify known properties, but 'plannedGradYear' does not exist in type 'StudentProfileInfo'. Did you mean to write 'PlannedGradYear'?
```

**Review by EjPlatzer — APPROVED** [review]
LGTM!

**Pull request merged**